crates/rpc/src/gas_price.rs

```diff
@@ -1,112 +1,93 @@
+                    tracing::debug!(from=?updated, "Using cached gas price value");
+                    return Some(gas_price);
+                }
+                _ => {
+                    tracing::debug!("Gas price missing or expired");
                 }
+            },
+            Err(reason) => {
+                tracing::debug!(%reason, "Failed to lock gas price mutex");
             }
+        }
 
-            // clear stale since it's not going to be useful for anyone
-            g.latest = None;
-
-            // this is an adaptation of https://fasterthanli.me/articles/request-coalescing-in-async-rust
-
-            if let Some(tx) = g.next.upgrade() {
-                // there's already an existing request being fulfilled
-                tx.subscribe()
-            } else {
-                let (tx, rx) = tokio::sync::broadcast::channel(1);
-
-                // the use of Weak works here, because the only strong reference is being sent to
-                // the async task, which upon completion holds the lock again while sending
-                // everyone listening the response, and clears the weak.
-                let tx = Arc::new(tx);
-
-                let inner = self.inner.clone();
-                let gateway = self.gateway.clone();
-
-                g.next = Arc::downgrade(&tx);
-
-                // Update the gas price from the starknet pending block.
-                tokio::spawn(async move {
-                    use starknet_gateway_types::reply::MaybePendingBlock;
-                    let gas_price = match gateway
-                        // Don't indefinitely retry as this could block the RPC request.
-                        .block_without_retry(pathfinder_common::BlockId::Pending)
-                        .await
-                    {
-                        Ok(b) => match b {
-                            MaybePendingBlock::Pending(b) => b.gas_price,
-                            MaybePendingBlock::Block(b) => match b.gas_price {
-                                Some(g) => g,
-                                None => {
-                                    tracing::debug!("Gas price missing in block");
-                                    let _ = tx.send(None);
-                                    return;
-                                }
-                            },
-                        },
-                        Err(reason) => {
-                            tracing::debug!(%reason, "Failed to fetch gas price");
-                            let _ = tx.send(None);
-                            return;
-                        }
-                    };
-
-                    let now = std::time::Instant::now();
-
-                    let gas_price = gas_price.0.into();
-
-                    let mut g = inner.lock().unwrap_or_else(|e| e.into_inner());
-                    g.latest.replace((now, gas_price));
+        if let Some(gas_price) = self.gas_price().await {
```

Comment:
This new code does not seem to do request coalescing anymore. I'd actually prefer keeping the previous logic that guaranteed that there will be at most one pending request at the same time.

Reply:
Shouldn't the outer lock be an async version?

Assuming that the cached value is outdated and the first caller triggered a get_block, the second caller to Cached::get will cause the thread of the second caller to block until get_block is done or times out. Which means all other futures in that second thread will also be blocked.

Reply:
> Shouldn't the outer lock be an async version?

I think it already is a `tokio::sync::Mutex`.

Reply:
> I think it already is a tokio::sync::Mutex.

You're right, my bad.